Web Inspector: harden `InspectorDebuggerAgent::didDispatchAsyncCall` to avoid problems if Web Inspector is opened after enqueuing before dispatch

dcrousso opened the pull request:
#### f407fbb287465a0ff68442eb7297862518e211cc
<pre>
Web Inspector: harden `InspectorDebuggerAgent::didDispatchAsyncCall` to avoid problems if Web Inspector is opened after enqueuing before dispatch
<a href="https://bugs.webkit.org/show_bug.cgi?id=242865">https://bugs.webkit.org/show_bug.cgi?id=242865</a>

Reviewed by Patrick Angle.

Imagine a series of events like this:

1. page calls `queueMicrotask`
2. developer opens Web Inspector
3. page calls `setTimeout`
4. microtask from (1) fires

Step (4) will cause the `AsyncStackTrace` created by step (3) to be immediately &quot;resolved&quot; since
`didDispatchAsyncCall` doesn&apos;t do any checking to validate that the async call matches the `AsyncCallType`
and identifier of the current `AsyncStackTrace`.

We should be providing the `AsyncCallType` and identifier at every step of the process so that we&apos;re
always &quot;resolving&quot; the right `AsyncStackTrace` for the given async call.

This will also give us some defense if in the future async calls somehow can become interleaved.

* Source/JavaScriptCore/inspector/agents/InspectorDebuggerAgent.h:
* Source/JavaScriptCore/inspector/agents/InspectorDebuggerAgent.cpp:
(Inspector::InspectorDebuggerAgent::willDispatchAsyncCall):
(Inspector::InspectorDebuggerAgent::didDispatchAsyncCall):
(Inspector::InspectorDebuggerAgent::didQueueMicrotask):
(Inspector::InspectorDebuggerAgent::willRunMicrotask):
(Inspector::InspectorDebuggerAgent::didRunMicrotask):
* Source/WebCore/dom/ScriptedAnimationController.cpp:
(WebCore::ScriptedAnimationController::serviceRequestAnimationFrameCallbacks):
* Source/WebCore/inspector/InspectorInstrumentation.h:
(WebCore::InspectorInstrumentation::didFireAnimationFrame):
* Source/WebCore/inspector/InspectorInstrumentation.cpp:
(WebCore::InspectorInstrumentation::didHandleEventImpl):
(WebCore::InspectorInstrumentation::didFireTimerImpl):
(WebCore::InspectorInstrumentation::didFireAnimationFrameImpl):
* Source/WebCore/inspector/agents/WebDebuggerAgent.h:
* Source/WebCore/inspector/agents/WebDebuggerAgent.cpp:
(WebCore::WebDebuggerAgent::willHandleEvent):
(WebCore::WebDebuggerAgent::didHandleEvent): Added.
(WebCore::WebDebuggerAgent::didDispatchPostMessage):
(WebCore::WebDebuggerAgent::didClearAsyncStackTraceData):
* Source/WebCore/inspector/agents/page/PageDebuggerAgent.h:
* Source/WebCore/inspector/agents/page/PageDebuggerAgent.cpp:
(WebCore::PageDebuggerAgent::didFireAnimationFrame): Added.

* Source/WTF/wtf/Vector.h:
(WTF::Malloc&gt;::removeLast): Added.
(WTF::Malloc&gt;::removeLastMatching): Added.
* Tools/TestWebKitAPI/Tests/WTF/Vector.cpp:
(TestWebKitAPI::TEST.WTF_Vector.RemoveLast): Added.
(TestWebKitAPI::TEST.WTF_Vector.RemoveLastMatching): Added.
New utility functions similar to `WTF::Vector::removeFirst` and `WTF::Vector::removeFirstMatching`
to make it easier to use a `WTF::Vector` like a LIFO structure (with a bit more flexibility).

Canonical link: <a href="https://commits.webkit.org/252869@main">https://commits.webkit.org/252869@main</a>
</pre>

patrickangle commented on Source/WTF/wtf/Vector.h:
```diff
@@ -1599,6 +1602,35 @@ inline bool Vector<T, inlineCapacity, OverflowHandler, minCapacity, Malloc>::rem
+template<typename MatchFunction>
+inline bool Vector<T, inlineCapacity, OverflowHandler, minCapacity, Malloc>::removeLastMatching(const MatchFunction& matches, size_t startIndex)
+{
+    for (size_t i = std::min(startIndex + 1, size()); i > 0; --i) {
+        if (matches(at(i - 1))) {
+            remove(i - 1);
```
It would probably be nice to refactor the `for` just a bit so that `i` is the actual index we are checking/removing for any given iteration e.g.
```
for (size_t i = std::min(startIndex, size() - 1); i >= 0; --i) {
    if (matches(at(i))) {
        remove(i);
```

dcrousso replied:
we cant do that because if `size()` is `0` then we get underflow :/

ditto with using `i >= 0` since when `i == 0` we'd still subtract one (from `--i`)

patrickangle replied:
doh. Good point. Unfortunate, but good point.

dcrousso committed:
Web Inspector: harden `InspectorDebuggerAgent::didDispatchAsyncCall` to avoid problems if Web Inspector is opened after enqueuing before dispatch
https://bugs.webkit.org/show_bug.cgi?id=242865

Reviewed by Patrick Angle.

Imagine a series of events like this:

1. page calls `queueMicrotask`
2. developer opens Web Inspector
3. page calls `setTimeout`
4. microtask from (1) fires

Step (4) will cause the `AsyncStackTrace` created by step (3) to be immediately "resolved" since
`didDispatchAsyncCall` doesn't do any checking to validate that the async call matches the `AsyncCallType`
and identifier of the current `AsyncStackTrace`.

We should be providing the `AsyncCallType` and identifier at every step of the process so that we're
always "resolving" the right `AsyncStackTrace` for the given async call.

This will also give us some defense if in the future async calls somehow can become interleaved.

* Source/JavaScriptCore/inspector/agents/InspectorDebuggerAgent.h:
* Source/JavaScriptCore/inspector/agents/InspectorDebuggerAgent.cpp:
(Inspector::InspectorDebuggerAgent::willDispatchAsyncCall):
(Inspector::InspectorDebuggerAgent::didDispatchAsyncCall):
(Inspector::InspectorDebuggerAgent::didQueueMicrotask):
(Inspector::InspectorDebuggerAgent::willRunMicrotask):
(Inspector::InspectorDebuggerAgent::didRunMicrotask):
* Source/WebCore/dom/ScriptedAnimationController.cpp:
(WebCore::ScriptedAnimationController::serviceRequestAnimationFrameCallbacks):
* Source/WebCore/inspector/InspectorInstrumentation.h:
(WebCore::InspectorInstrumentation::didFireAnimationFrame):
* Source/WebCore/inspector/InspectorInstrumentation.cpp:
(WebCore::InspectorInstrumentation::didHandleEventImpl):
(WebCore::InspectorInstrumentation::didFireTimerImpl):
(WebCore::InspectorInstrumentation::didFireAnimationFrameImpl):
* Source/WebCore/inspector/agents/WebDebuggerAgent.h:
* Source/WebCore/inspector/agents/WebDebuggerAgent.cpp:
(WebCore::WebDebuggerAgent::willHandleEvent):
(WebCore::WebDebuggerAgent::didHandleEvent): Added.
(WebCore::WebDebuggerAgent::didDispatchPostMessage):
(WebCore::WebDebuggerAgent::didClearAsyncStackTraceData):
* Source/WebCore/inspector/agents/page/PageDebuggerAgent.h:
* Source/WebCore/inspector/agents/page/PageDebuggerAgent.cpp:
(WebCore::PageDebuggerAgent::didFireAnimationFrame): Added.

* Source/WTF/wtf/Vector.h:
(WTF::Malloc>::removeLast): Added.
(WTF::Malloc>::removeLastMatching): Added.
* Tools/TestWebKitAPI/Tests/WTF/Vector.cpp:
(TestWebKitAPI::TEST.WTF_Vector.RemoveLast): Added.
(TestWebKitAPI::TEST.WTF_Vector.RemoveLastMatching): Added.
New utility functions similar to `WTF::Vector::removeFirst` and `WTF::Vector::removeFirstMatching`
to make it easier to use a `WTF::Vector` like a LIFO structure (with a bit more flexibility).

Canonical link: https://commits.webkit.org/252869@main
